File: src/input_normalizer/input_normalize.py
from datetime import datetime
from typing import Any, Dict, List


def normalize_input(raw: Dict[str, Any]) -> Dict[str, Any]:
    """
    Normalize user-provided flight search JSON into canonical structure
    matching FlightSearchRequest (as defined in the updated .proto schema).
    """

    def parse_date(date_str: Any) -> str:
        """Normalize various date formats to 'YYYY-MM-DD'."""
        if not date_str:
            return None
        try:
            return datetime.fromisoformat(str(date_str)).strftime("%Y-%m-%d")
        except ValueError:
            for fmt in ("%Y/%m/%d", "%m/%d/%Y", "%Y-%m-%d"):
                try:
                    return datetime.strptime(str(date_str), fmt).strftime("%Y-%m-%d")
                except ValueError:
                    continue
        raise ValueError(f"Invalid date format: {date_str}")

    # --- Extract core fields ---
    origin = raw.get("origin")
    destination = raw.get("destination")

    if not origin or not destination:
        raise ValueError("Missing required origin or destination airport code.")

    departure_date = parse_date(raw.get("departure_date"))
    return_date = parse_date(raw.get("return_date"))

    # --- Infer trip type ---
    if "trip" in raw:
        trip_type = raw["trip"].replace("-", "_").lower()
    elif return_date:
        trip_type = "trip_type_round"
    else:
        trip_type = "trip_type_one_way"

    # --- Cabin class (seat) ---
    cabin_class = (raw.get("seat") or raw.get("cabin_class") or "economy").lower()

    # --- Passengers (defaults applied) ---
    pax = raw.get("passengers", {})
    passengers = {
        "adult": pax.get("adult", 1),
        "child": pax.get("child", 0),
        "infant": pax.get("infant", 0),
    }

    # --- Extract and validate fixed flights ---
    fixed_flights = raw.get("fixed_flights", {})
    outbound_segments = []
    return_segments = []

    if isinstance(fixed_flights, dict):
        outbound_segments = fixed_flights.get("outbound", [])
        return_segments = fixed_flights.get("return", [])
    else:
        raise ValueError("fixed_flights must be a dict if provided.")

    if return_segments and not outbound_segments:
        raise ValueError("Invalid input: 'return' segments cannot exist without 'outbound'.")

    def normalize_segment(segment: Dict[str, Any]) -> Dict[str, Any]:
        """Convert flight segment into canonical form."""
        return {
            "origin_airport": segment.get("origin_airport"),
            "destination_airport": segment.get("destination_airport"),
            "departure_date": parse_date(segment.get("date")),
            "airline_code": segment.get("airline_code"),
            "flight_number": segment.get("flight_number"),
        }

    itinerary: List[Dict[str, Any]] = []

    # --- Build outbound leg ---
    outbound_leg = {
        "travel_date": departure_date,
        "segments": [normalize_segment(f) for f in outbound_segments if f],
        "origin": {"code": origin},
        "destination": {"code": destination},
    }
    itinerary.append(outbound_leg)

    # --- Build return leg (if applicable) ---
    if return_date:
        itinerary.append({
            "travel_date": return_date,
            "segments": [normalize_segment(f) for f in return_segments if f],
            "origin": {"code": destination},
            "destination": {"code": origin},
        })

    # --- Final normalized structure ---
    canonical = {
        "itinerary": itinerary,
        "cabin_class": cabin_class,
        "passengers": passengers,
        "trip_type": trip_type,
    }

    return canonical


# --- Inline testing block ---
if __name__ == "__main__":
    import sys
    import json
    
    input_json = sys.stdin.read()
    input_data = json.loads(input_json)
    normalized_data = normalize_input(input_data)
    print(json.dumps(normalized_data))

File: src/input_normalizer/input_normalize.js
export function normalizeInput(raw) {
  function parseDate(dateStr) {
    if (!dateStr) return null;
    const s = String(dateStr).trim();

    // YYYY-MM-DD or YYYY/MM/DD
    const isoMatch = s.match(/^(\d{4})[-/](\d{2})[-/](\d{2})$/);
    if (isoMatch) return `${isoMatch[1]}-${isoMatch[2]}-${isoMatch[3]}`;

    // MM/DD/YYYY
    const usMatch = s.match(/^(\d{2})\/(\d{2})\/(\d{4})$/);
    if (usMatch) return `${usMatch[3]}-${usMatch[1]}-${usMatch[2]}`;

    // ISO with time component
    if (s.includes("T")) return s.split("T")[0];

    throw new Error(`Invalid date format: ${dateStr}`);
  }

  const origin = raw.origin;
  const destination = raw.destination;
  if (!origin || !destination) {
    throw new Error("Missing required origin or destination airport code.");
  }

  const today = new Date();
  today.setUTCHours(0, 0, 0, 0);

  let departureDateStr = parseDate(raw.departure_date);
  let returnDateStr = parseDate(raw.return_date);

  if (!departureDateStr) {
    const d = new Date(today);
    d.setUTCDate(d.getUTCDate() + 30);
    departureDateStr = d.toISOString().split("T")[0];
  } else {
    const depDt = new Date(departureDateStr + "T00:00:00Z");
    if (depDt < today) {
      const d = new Date(today);
      d.setUTCDate(d.getUTCDate() + 30);
      departureDateStr = d.toISOString().split("T")[0];
    }
  }

  if (returnDateStr) {
    const retDt = new Date(returnDateStr + "T00:00:00Z");
    const depDt = new Date(departureDateStr + "T00:00:00Z");
    if (retDt < depDt) {
      const d = new Date(depDt);
      d.setUTCDate(d.getUTCDate() + 7);
      returnDateStr = d.toISOString().split("T")[0];
    }
  }

  let tripType;
  if (raw.trip) {
    tripType = raw.trip.replace(/-/g, "_").toLowerCase();
  } else if (returnDateStr) {
    tripType = "trip_type_round";
  } else {
    tripType = "trip_type_one_way";
  }

  const cabinClass = (raw.seat || raw.cabin_class || "economy").toLowerCase();

  const pax = raw.passengers || {};
  const passengers = {
    adult: pax.adult ?? 1,
    child: pax.child ?? 0,
    infant: pax.infant ?? 0,
  };

  const fixedFlights = raw.fixed_flights || {};
  if (typeof fixedFlights !== "object" || Array.isArray(fixedFlights)) {
    throw new Error("fixed_flights must be a dict if provided.");
  }
  const outboundSegments = fixedFlights.outbound || [];
  const returnSegments = fixedFlights.return || [];
  if (returnSegments.length > 0 && outboundSegments.length === 0) {
    throw new Error("Invalid input: 'return' segments cannot exist without 'outbound'.");
  }

  function normalizeSegment(segment) {
    return {
      origin_airport: segment.origin_airport || segment.origin,
      destination_airport: segment.destination_airport || segment.destination,
      departure_date: parseDate(segment.date || segment.flight_date || segment.departure_date),
      airline_code: segment.airline_code,
      flight_number: segment.flight_number,
    };
  }

  const itinerary = [
    {
      travel_date: departureDateStr,
      segments: outboundSegments.filter(Boolean).map(normalizeSegment),
      origin: { code: origin },
      destination: { code: destination },
    },
  ];

  if (returnDateStr) {
    itinerary.push({
      travel_date: returnDateStr,
      segments: returnSegments.filter(Boolean).map(normalizeSegment),
      origin: { code: destination },
      destination: { code: origin },
    });
  }

  return {
    itinerary,
    cabin_class: cabinClass,
    passengers,
    trip_type: tripType,
  };
}
